src/grpc/qgrpcchannel.cpp

```diff
@@ -0,0 +1,343 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2019 Giulio Girardi <giulio.girardi@protechgroup.it>
+ *
+ * This file is part of QtProtobuf project https://git.semlanik.org/semlanik/qtprotobuf
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the "Software"), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify,
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies
+ * or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "qgrpcchannel.h"
+#include "qgrpcchannel_p.h"
+
+#include <QEventLoop>
+#include <QThread>
+
+#include <memory>
+#include <thread>
+#include <unordered_map>
+
+#include <grpcpp/generic/generic_stub.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/impl/codegen/byte_buffer.h>
+#include <grpcpp/impl/codegen/channel_interface.h>
+#include <grpcpp/impl/codegen/client_callback_impl.h>
+#include <grpcpp/impl/codegen/client_context.h>
+#include <grpcpp/impl/codegen/client_unary_call.h>
+#include <grpcpp/impl/codegen/completion_queue.h>
+#include <grpcpp/impl/codegen/rpc_method.h>
+#include <grpcpp/impl/codegen/slice.h>
+#include <grpcpp/impl/codegen/status.h>
+#include <grpcpp/impl/codegen/status_code_enum.h>
+#include <grpcpp/impl/codegen/sync_stream_impl.h>
+#include <grpcpp/security/credentials.h>
+
+#include "qabstractgrpccredentials.h"
+#include "qgrpcasyncreply.h"
+#include "qgrpcstatus.h"
+#include "qgrpcsubscription.h"
+#include "qabstractgrpcclient.h"
+#include "qgrpccredentials.h"
+#include "qprotobufserializerregistry_p.h"
+#include "qtprotobuflogging.h"
+
+using namespace QtProtobuf;
+
+namespace QtProtobuf {
+
+static inline grpc::Status parseByteBuffer(const grpc::ByteBuffer &buffer, QByteArray *data) {
+    std::vector<grpc::Slice> slices;
+    auto status = buffer.Dump(&slices);
+
+    for (auto slice : slices) {
+        data->append(QByteArray((const char *)slice.begin(), slice.size()));
+    }
+
+    return status;
+}
+
+static inline grpc::Status parseQByteArray(const QByteArray &bytearray, grpc::ByteBuffer *buffer) {
```

Comment:
Probably non-constant reference to grpc:ByteBuffer &buffer is better in this case?